File: 2022/rust/src/day2.rs
use std::collections::HashMap;

fn part1(input: String) -> i32 {
    let my_strategy = HashMap::from([
        ("X", ShapePlay::Rock),
        ("Y", ShapePlay::Paper),
        ("Z", ShapePlay::Scissors),
    ]);
    let opponent_strategy = HashMap::from([
        ("A", ShapePlay::Rock),
        ("B", ShapePlay::Paper),
        ("C", ShapePlay::Scissors),
    ]);

    let mut my_score = 0;
    for line in input.split("\n") {
        let inputs: Vec<&str> = line.split(" ").collect();
        if inputs.len() < 2 {
            continue;
        }

        let my_input = my_strategy.get(inputs[1]).unwrap();
        let opponent_input = opponent_strategy.get(inputs[0]).unwrap();

        my_score += my_input.score_against_opponent(opponent_input);
    }

    return my_score;
}

fn part2(input: String) -> i32 {
    let my_strategy = HashMap::from([
        ("X", PlayDecision::Lose),
        ("Y", PlayDecision::Draw),
        ("Z", PlayDecision::Win),
    ]);
    let opponent_strategy = HashMap::from([
        ("A", ShapePlay::Rock),
        ("B", ShapePlay::Paper),
        ("C", ShapePlay::Scissors),
    ]);

    let mut my_score = 0;
    for line in input.split("\n") {
        let inputs: Vec<&str> = line.split(" ").collect();
        if inputs.len() < 2 {
            continue;
        }

        let opponent_input = opponent_strategy.get(inputs[0]).unwrap();
        let my_input = my_strategy
            .get(inputs[1])
            .unwrap()
            .shape_against_opponent(opponent_input);

        my_score += my_input.score_against_opponent(opponent_input);
    }

    return my_score;
}

#[derive(PartialEq)]
enum PlayDecision {
    Win,
    Lose,
    Draw,
}

impl PlayDecision {
    fn score(&self) -> i32 {
        match self {
            PlayDecision::Win => 6,
            PlayDecision::Lose => 0,
            PlayDecision::Draw => 3,
        }
    }

    fn shape_against_opponent(&self, opponent: &ShapePlay) -> ShapePlay {
        if self == &PlayDecision::Win {
            match opponent {
                ShapePlay::Rock => return ShapePlay::Paper,
                ShapePlay::Paper => return ShapePlay::Scissors,
                ShapePlay::Scissors => return ShapePlay::Rock,
            }
        } else if self == &PlayDecision::Lose {
            match opponent {
                ShapePlay::Rock => return ShapePlay::Scissors,
                ShapePlay::Paper => return ShapePlay::Rock,
                ShapePlay::Scissors => return ShapePlay::Paper,
            }
        }

        return opponent.clone();
    }
}

#[derive(Clone, PartialEq)]
enum ShapePlay {
    Rock,
    Paper,
    Scissors,
}

impl ShapePlay {
    fn score(&self) -> i32 {
        match self {
            ShapePlay::Rock => 1,
            ShapePlay::Paper => 2,
            ShapePlay::Scissors => 3,
        }
    }

    fn score_against_opponent(&self, opponent_input: &ShapePlay) -> i32 {
        let decision: PlayDecision;
        if self == opponent_input {
            decision = PlayDecision::Draw;
        } else if self == &ShapePlay::Rock {
            match opponent_input {
                ShapePlay::Scissors => decision = PlayDecision::Win,
                ShapePlay::Paper => decision = PlayDecision::Lose,
                _ => panic!("should have been handled"),
            }
        } else if self == &ShapePlay::Paper {
            match opponent_input {
                ShapePlay::Rock => decision = PlayDecision::Win,
                ShapePlay::Scissors => decision = PlayDecision::Lose,
                _ => panic!("should have been handled"),
            }
        } else {
            match opponent_input {
                ShapePlay::Rock => decision = PlayDecision::Lose,
                ShapePlay::Paper => decision = PlayDecision::Win,
                _ => panic!("should have been handled"),
            }
        }

        return decision.score() + self.score();
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    const FILE_PATH: &'static str = "../../Packages/AOC2022/Sources/AOC2022/Inputs/day2.txt";

    #[test]
    fn part1_works() {
        let input = fs::read_to_string(FILE_PATH).expect("Could not read file");
        let result = super::part1(input);
        assert_eq!(result, 14375);
    }

    #[test]
    fn part2_works() {
        let input = fs::read_to_string(FILE_PATH).expect("Could not read file");
        let result = super::part2(input);
        assert_eq!(result, 10274);
    }
}

File: Swift/AOC2022/Sources/AOC2022/Day2.swift
//
//  Day2.swift
//  
//
//  Created by Kamaal M Farah on 01/12/2022.
//

import Foundation
import ShrimpExtensions

private let OPPONENT_STRATEGY: [String: ShapePlay] = [
    "A": .rock,
    "B": .paper,
    "C": .scissors,
]

extension AOC2022 {
    public struct Day2: DayScaffold {
        public static let resourceName = "day2"

        private init() { }

        public struct Part1 {
            private init() { }

            public static func excecute(with input: String) -> Int {
                let myStrategy: [String: ShapePlay] = [
                    "X": .rock,
                    "Y": .paper,
                    "Z": .scissors,
                ]

                var myTotalScore = 0
                for line in input.splitLines {
                    let inputs = line.split(separator: " ").map({ String($0) })
                    let opponentInput = OPPONENT_STRATEGY[inputs[0]]!
                    let myInput = myStrategy[inputs[1]]!

                    myTotalScore += myInput.scoreAgainstOpponent(opponentInput)
                }

                return myTotalScore
            }
        }

        public struct Part2 {
            private init() { }

            public static func excecute(with input: String) -> Int {
                let myStrategy: [String: PlayDecision] = [
                    "X": .lose,
                    "Y": .draw,
                    "Z": .win,
                ]

                var myTotalScore = 0
                for line in input.splitLines {
                    let inputs = line.split(separator: " ").map({ String($0) })
                    let opponentInput = OPPONENT_STRATEGY[inputs[0]]!
                    let myInput = myStrategy[inputs[1]]!.shapeAgainsOpponent(opponentInput)

                    myTotalScore += myInput.scoreAgainstOpponent(opponentInput)
                }

                return myTotalScore
            }
        }
    }
}

private enum PlayDecision: Int {
    case win = 6
    case lose = 0
    case draw = 3

    var score: Int {
        rawValue
    }

    func shapeAgainsOpponent(_ opponent: ShapePlay) -> ShapePlay {
        if self == .win {
            switch opponent {
            case .rock:
                return .paper
            case .paper:
                return .scissors
            case .scissors:
                return .rock
            }
        }

        if self == .lose {
            switch opponent {
            case .rock:
                return .scissors
            case .paper:
                return .rock
            case .scissors:
                return .paper
            }
        }

        return opponent
    }
}

private enum ShapePlay: Int {
    case rock = 1
    case paper = 2
    case scissors = 3

    var score: Int {
        rawValue
    }

    func scoreAgainstOpponent(_ opponent: ShapePlay) -> Int {
        let ownScore = score

        let decision: PlayDecision
        if self == opponent {
            decision = .draw
        } else if self == .rock {
            switch opponent {
            case .scissors:
                decision = .win
            case .paper:
                decision = .lose
            default:
                fatalError("should have been handled")
            }
        } else if self == .paper {
            switch opponent {
            case .rock:
                decision = .win
            case .scissors:
                decision = .lose
            default:
                fatalError("should have been handled")
            }
        } else {
            switch opponent { // scissors
            case .rock:
                decision = .lose
            case .paper:
                decision = .win
            default:
                fatalError("should have been handled")
            }
        }

        return ownScore + decision.score
    }
}
